How to use GPUs

pt247 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://nebari.dev/community
-->

## Reference Issues or PRs
Closes #119
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create a link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [x] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

## Documentation

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Nebari Internals docs.
  - If no docs exist:
    - Create a stub for documentation, including bullet points for how to use the feature, code snippets (including from
      happy path tests), etc.
We want to ensure that the content produced for Nebari is accessible; if you are making significant contributions,
please address the access-centred guidelines in your content and complete our checklist. Thanks to @isabela-pf for this checklist :)
-->

### Access-centered content checklist

#### Text styling

- [x] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [x] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [x] This content adheres to the Nebari style guides.

#### Non-text content

- [x] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [x] If the content were to be read as plain text, it still makes sense, and no information is missing.

## Any other comments?

<!--
Please be aware that we are a loose team of volunteers, so patience is necessary;
assistance handling other issues is very welcome.
We value all user contributions. If we are slow to review, either the pull request needs some benchmarking, tinkering,
convincing, etc., or the reviewers are likely busy. In either case,
we ask for your understanding during the
review process.
Thanks for contributing to Nebari 🙏🏼!
-->

ericdatakelly commented:
If you add just a little more info, you could remove Best Practices for PyTorch on Nebari.  Also, I think that info is slightly out of date. `pytorch-gpu` is now `pytorch-cuda` as shown by "Run this command" from pytorch.org matrix.  Also for the conda yaml config, I suggest including the channels recommended by the pytorch matrix so the user has a complete example.  They need to be above conda-forge.

```yaml
channels:
  - pytorch
  - nvidia
  - conda-forge
dependencies:
  - pytorch
  - ipykernel
variables:
  CONDA_OVERRIDE_CUDA: "12.1"
```

Have you tested each method (2.1, 2.2) to be sure they work and you are describing all steps?

pt247 commented:
> If you add just a little more info, you could remove Best Practices for PyTorch on Nebari. Also, I think that info is slightly out of date. `pytorch-gpu` is now `pytorch-cuda` as shown by "Run this command" from pytorch.org matrix. Also for the conda yaml config, I suggest including the channels recommended by the pytorch matrix so the user has a complete example. They need to be above conda-forge.

Agreed, I have updated it now. 

> Have you tested each method (2.1, 2.2) to be sure they work and you are describing all steps?

I have now tested and updated the document.

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+
+# Using GPUs on Nebari
+## Introduction
+Provide a complete overview of using GPUs on Nebari, including server setup, environment setup, and validation.
```
```suggestion
Overview of using GPUs on Nebari including server setup, environment setup, and validation.
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+---
+id: use-gpus
+title: Using GPUs on Nebari
+description: Provide a complete overview of using GPUs on Nebari, including server setup, environment setup, and validation.
```
```suggestion
description: Overview of using GPUs on Nebari including server setup, environment setup, and validation.
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+3. Run the command `nvidia-smi`. The top right corner of the command's output should have the highest supported driver.
+    ![nvidia-smi-output](/img/how-tos/nvidia-smi-output.png)
+
+    If you get `nvidia-smi: command not found`. You are most likely on a non-GPU server. Shutdown your server, and start up a GPU-enabled server.
```
```suggestion
    If you get the error `nvidia-smi: command not found`, you are most likely on a non-GPU server. Shutdown your server, and start up a GPU-enabled server.
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+By default, `conda-store` will build CPU-compatible packages. To build GPU-compatible packages, we the following.
+### Build a GPU-compatible environments
+By default, conda-store will build CPU-compatible packages. To build GPU-compatible packages, we have a couple of options:
+1. **Creating config using PyTorch documentation**:
```
We need to swap these two sections (1 & 2). We want approach 2 to be the way folks do it, but just make them aware that approach 1 also exists.

kcpevey replied:
```suggestion
1. **Create the environment specification using the CUDA-specific packages as shown in the PyTorch documentation configuration (alternative approach)**:
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+
+By default, `conda-store` will build CPU-compatible packages. To build GPU-compatible packages, we the following.
+### Build a GPU-compatible environments
+By default, conda-store will build CPU-compatible packages. To build GPU-compatible packages, we have a couple of options:
```
```suggestion
By default, `conda-store` will build CPU-compatible packages. To build GPU-compatible packages, we have two options:
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,105 @@
+    ```
+    You can use **GUI <-> YAML** Toggle to edit the config.
+
+2. **Build a GPU-compatible environment in conda-store using `CONDA_OVERRIDE_CUDA`**:
````
```suggestion
2. **Create environment specification using `CONDA_OVERRIDE_CUDA` (recommended approach)**:
```

pt247 replied:
Changed.

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+    ![nvidia-smi-output](/img/how-tos/nvidia-smi-output.png)
+
+    If you get `nvidia-smi: command not found`. You are most likely on a non-GPU server. Shutdown your server, and start up a GPU-enabled server.
+
```
```suggestion
**Compatible environments for this server must contain CUDA versions *below* the GPU server version. For example, the server in this case is on 12.4. All environments used on this server must contain packages build with CUDA<=12.4.**

```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+By default, conda-store will build CPU-compatible packages. To build GPU-compatible packages, we have a couple of options:
+1. **Creating config using PyTorch documentation**:
+
+    You will need to check PyTorch documentation to [select a supported version of CUDA from the support matrix](https://pytorch.org/get-started/locally/) to determine the latest override version for `CONDA_OVERRIDE_CUDA`.
```
```suggestion
    You can check [PyTorch documentation](https://pytorch.org/get-started/locally/) to get a quick list of the necessary CUDA-specific packages.
```

pt247 replied:
Changed.

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+1. **Creating config using PyTorch documentation**:
+
+    You will need to check PyTorch documentation to [select a supported version of CUDA from the support matrix](https://pytorch.org/get-started/locally/) to determine the latest override version for `CONDA_OVERRIDE_CUDA`.
+    Select the following options to get the latest CUDA version:
+    - PyTorch Build = Stable
+    - Your OS          = Linux
+    - Package          = Conda
+    - Language         = Python
+    - Compute Platform = 12.1 (Choose the latest supported platform)
```
```suggestion
At the time of writing, 12.1 was the latest supported version. Select the version that matches the `nvidia-smi` output (see above) on your server. 
```

kcpevey replied:
My point was more that you actually DONT want to get the latest CUDA from that interface - you want the CUDA version that matches the GPU that you're running (or less).

ericdatakelly replied:
Actually I was also making a different point. :) 

I want to be sure a person who uses a screen reader can get the information if the list is removed.  I also want to help the user understand what choices to make when using the matrix.  For example, is it ok to use LibTorch on Nebari?  Let's tell them to choose linux, conda, and python.

As far as cuda version, the user will want to know which version of cuda is closest to the one given by `nvidia-smi`, which is unknown until running the matrix.  For example, only 12.1 and lower are compatible right now, so if the user tries to install 12.3, it will fail.

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,105 @@
+    ```
+    conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
+    ```
+    The corresponding yaml config is:
````
```suggestion
    The corresponding yaml config would be:
```

kcpevey reviewed (changes requested):
I think at this point, we can remove https://deploy-preview-471--nebari-docs.netlify.app/docs/how-tos/pytorch-best-practices from the docs. @ericdatakelly is there anything else we need to include here to ensure coverage of the same material?

ericdatakelly commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,105 @@
+    conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
+    ```
+    The corresponding yaml config is:
+    ```
````
It's not necessary to use the yaml editor.  The cuda options can be chosen in the gui.  But this is an easy way to show the packages chosen, so I think it's ok.

pt247 replied:
You are right. It's easier to write docs this way. I will leave it as such.

ericdatakelly commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,105 @@
+      - ipykernel
+    variables: {}
+    ```
+    You can use **GUI <-> YAML** Toggle to edit the config.
````
```suggestion
    Note that the order of the channels is respected by conda, so keep pytorch at the top, then nvidia, then conda-forge. You can use **GUI <-> YAML** Toggle to edit the config.
```

pt247 replied:
Updated.

ericdatakelly commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,105 @@
+      CONDA_OVERRIDE_CUDA: "12.1"
+    ```
+
+    Replace this section with an override for `CONDA_OVERRIDE_CUDA` to tell the conda-store to build a GPU-compatible environment instead of the default CPU-compatible one.
````
```suggestion
    Add the `CONDA_OVERRIDE_CUDA` override to the variables section to tell conda-store to build a GPU-compatible environment.
```

ericdatakelly commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,105 @@
+    ```
+
+    Replace this section with an override for `CONDA_OVERRIDE_CUDA` to tell the conda-store to build a GPU-compatible environment instead of the default CPU-compatible one.
+    ![pytorch-cuda-override.png](/img/how-tos/pytorch-cuda-override.png)
````
The image is redundant with the yaml, so let's remove it.
```suggestion
```

pt247 replied:
Done.

ericdatakelly commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+
+    Replace this section with an override for `CONDA_OVERRIDE_CUDA` to tell the conda-store to build a GPU-compatible environment instead of the default CPU-compatible one.
+    ![pytorch-cuda-override.png](/img/how-tos/pytorch-cuda-override.png)
+    :::note
```
Be sure to only keep the first mention of the max cuda version (after the two sections are swapped).

ericdatakelly commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,105 @@
+
+
+## 3. Validating the setup
+You can validate your setup and ensure that your GPU server is compatible with your GPU-compatible environment with the following process. Open a Jupyter Notebook and load the environment from above. Run the following commands to validate your setup:
```
```suggestion
You can check that your GPU server is compatible with your conda environment by opening a Jupyter Noteboo, loading the environment, and running the following code:
```

ericdatakelly reviewed (commented):
I added some suggestions.

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,111 @@
+---
+id: use-gpus
+title: Using GPUs on Nebari
```
```suggestion
title: Use GPUs on Nebari
```
This keeps it grammatically correct to the other titles.

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,111 @@
+    Alternatively, you can configure the same config using UI.
+
+    Add the `CONDA_OVERRIDE_CUDA` override to the variables section to tell conda-store to build a GPU-compatible environment.
+    :::note
```
```suggestion

    :::note
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,111 @@
+    ```
+    :::note
+    The order of the channels is respected by conda, so keep pytorch at the top, then nvidia, then conda-forge.
+    note
````
```suggestion
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,111 @@
+
+    Add the `CONDA_OVERRIDE_CUDA` override to the variables section to tell conda-store to build a GPU-compatible environment.
+
+    :::note
+    - At the time of writing this document, the latest CUDA version was showing as `12.1`. Please follow the above step to determine the latest override value for the `CONDA_OVERRIDE_CUDA` environment variable.
+    - Please ensure that your choice from PyTorch documentation is not greater than the highest supported version in the `nvidia-smi` output (captured above).
+    :::
```
```suggestion
:::note
    - At the time of writing this document, the latest CUDA version was showing as `12.1`. Please follow the above step to determine the latest override value for the `CONDA_OVERRIDE_CUDA` environment variable.
    - Please ensure that your choice from PyTorch documentation is not greater than the highest supported version in the `nvidia-smi` output (captured above).
:::
```

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
```diff
@@ -0,0 +1,111 @@
+    Add the `CONDA_OVERRIDE_CUDA` override to the variables section to tell conda-store to build a GPU-compatible environment.
+
+:::note
+    - At the time of writing this document, the latest CUDA version was showing as `12.1`. Please follow the above step to determine the latest override value for the `CONDA_OVERRIDE_CUDA` environment variable.
+    - Please ensure that your choice from PyTorch documentation is not greater than the highest supported version in the `nvidia-smi` output (captured above).
```
```suggestion
At the time of writing this document, the latest CUDA version was showing as `12.1`. Please follow the above step to determine the latest override value for the `CONDA_OVERRIDE_CUDA` environment variable.

Please ensure that your choice from PyTorch documentation is not greater than the highest supported version in the `nvidia-smi` output (captured above).
```

kcpevey reviewed (approved):
This looks good to me! @ericdatakelly do you want to take a final look before we merge?

ericdatakelly reviewed (commented):
I found some minor typos and suggested a few other changes, but hopefully they are easy and can just be accepted or rejected.  I do not need to review again - it looks great and I trust your decisions on my suggestions.  I also see that this image is not used anywhere, but I did not feel it was needed: docs/static/img/how-tos/conda-store-yaml-toggle.png.

kcpevey commented:
Thanks @ericdatakelly for your review!

@pt247 can you remove this file? docs/static/img/how-tos/conda-store-yaml-toggle.png

kcpevey commented on docs/docs/how-tos/use-gpus.mdx:
````diff
@@ -0,0 +1,109 @@
+    While creating a new config, click on the **GUI <-> YAML** Toggle to edit yaml config.
+    ```
+    channels:
+      - nvidia
````
```suggestion
```

kcpevey commented:
>Why remove pytorch and nvidia channels? The pytorch matrix tells you where the packages are, so you should include them. Conda forge sometimes does not have the latest packages.

I think I am torn here between doing what the pytorch website tells you vs what folks who use this for work tell us (which is, pull everything from conda-forge if you can, sometimes conda-forge and pytorch channels have conflicting builds)

kcpevey commented:
We are going to leave pytorch in the channel list since there are still some pytorch related packages that are not yet on conda-forge. We'll leave both pytorch and nvidia in the "follow the pytorch docs" since this is what they suggest.
